### PR TITLE
fix rsa encrypt large msg bug

### LINKED ISSUE #39: IPNS Publishing: crypto/rsa: message too long for RSA public key size
see comment [#5536](https://github.com/ipfs/go-ipfs/issues/5536)

https://github.com/libp2p/go-libp2p-crypto/blob/137c71cdda7aadcaad5eaf03301f7c4f173858e2/rsa.go#L45-L47

### PR DESCRIPTION
**Pull request by kjzz**
fix #39 and [go-ipfs #5536](https://github.com/ipfs/go-ipfs/issues/5536)

### CODE REVIEW

File: rsa.go [hunk 1]
```diff
@@ -9,7 +9,8 @@ import (
 
 	pb "github.com/libp2p/go-libp2p-crypto/pb"
 
-	sha256 "github.com/minio/sha256-simd"
+	"bytes"
+	"github.com/minio/sha256-simd"
```
Comment (vyzo) [inline rsa.go]: please don't change that.

File: rsa.go [hunk 2]
```diff
@@ -43,7 +44,17 @@ func (pk *RsaPublicKey) Raw() ([]byte, error) {
 }
 
 func (pk *RsaPublicKey) Encrypt(b []byte) ([]byte, error) {
-	return rsa.EncryptPKCS1v15(rand.Reader, pk.k, b)
+	partLen := pk.k.N.BitLen()/8 - 11
+	chunks := split([]byte(b), partLen)
+	buffer := bytes.NewBufferString("")
```
Comment (vyzo) [inline rsa.go]: why a buffer string? We want raw bytes.

**Comment by kjzz** [comment]
Thx @vyzo , i have update my pr.please review again

File: rsa.go [hunk 3]
```diff
@@ -9,6 +9,7 @@ import (
 
 	pb "github.com/libp2p/go-libp2p-crypto/pb"
 
+	"bytes"
```
Comment (vyzo) [inline rsa.go]: we put system imports on top, all grouped together.

File: rsa.go [hunk 4]
```diff
@@ -43,7 +44,17 @@ func (pk *RsaPublicKey) Raw() ([]byte, error) {
 }
 
 func (pk *RsaPublicKey) Encrypt(b []byte) ([]byte, error) {
-	return rsa.EncryptPKCS1v15(rand.Reader, pk.k, b)
+	partLen := pk.k.N.BitLen()/8 - 11
```
Comment (vyzo) [inline rsa.go]: why the -11?

File: rsa.go [hunk 5]
```diff
@@ -43,7 +44,17 @@ func (pk *RsaPublicKey) Raw() ([]byte, error) {
...
+	chunks := split([]byte(b), partLen)
```
Comment (vyzo) [inline rsa.go]: why the cast to `[]byte`? It's already of that type.

File: rsa.go [hunk 6]
```diff
@@ -43,7 +44,17 @@ func (pk *RsaPublicKey) Raw() ([]byte, error) {
...
+	buffer := bytes.NewBuffer([]byte(""))
```
Comment (vyzo) [inline rsa.go]: that's incorrect usage of the buffer -- just pass nil or create a new buffer object directly.
See the docs in https://godoc.org/bytes

File: rsa.go [hunk 7]
```diff
@@ -43,7 +44,17 @@ func (pk *RsaPublicKey) Raw() ([]byte, error) {
...
+	for _, chunk := range chunks {
+		bytes, err := rsa.EncryptPKCS1v15(rand.Reader, pk.k, chunk)
```
Comment (vyzo) [inline rsa.go]: this is not gonna cut it, you make chunks and apply padding. That's not the way to encrypt large messages with RSA -- you typically create a symmetric encryption key, encrypt that with RSA, and use it to encrypt the message.

Comment (kjzz) [reply rsa.go]: The reason why i cut message into chunk is if we use `EncryptPKCS1v15` to encrypt message,The message must be no longer than the length of the public modulus minus 11 bytes.

You can see the note above the function:
https://github.com/golang/go/blob/bf8e6b70276ddfd3e070964542020fe782c1eec8/src/crypto/rsa/pkcs1v15.go#L29-L39

And it has caused bug in [go-ipfs #5536](https://github.com/ipfs/go-ipfs/issues/5536).

File: rsa.go (hunk 5 above)
Comment (kjzz) [reply rsa.go]: if we use `EncryptPKCS1v15` to encrypt message,the message must be no longer than the length of the public modulus minus 11 bytes.

File: rsa.go (hunk 4 above)
Comment (kjzz) [reply rsa.go]: if we use EncryptPKCS1v15 to encrypt message,The message must be no longer than the length of the public modulus minus 11 bytes.So i should `-11`.

Comment (kjzz) [reply rsa.go]: if we use `EncryptPKCS1v15` to encrypt message,The message must be no longer than the length of the public modulus minus 11 bytes.So i need to `-11`.

**Comment by Stebalien** [comment]
This function is actually unused. We should remove it.

We *initially* supported encryption but realized we should have separate encryption/signature keys so we removed that from the interface.

**Comment by Stebalien** [comment]
Actually, let's just leave these for now. They should (technically) work and users will (usually) just stick with the functions on the interfaces anyways.

But yeah, as @vyzo said, the correct way to do this is to encrypt the message with a block-cipher using a symmetric key and then encrypt the symmetric key with the public key.

**Pull request closed without merging**